beta-operator-rewards/rewards-requirements.ts

```diff
@@ -0,0 +1,478 @@
+      postEvents,
+      stakingProvider
+    );
+
+    return await authorizationPostRewardsInterval(
+      postIntervalEvents,
+      application,
+      stakingProvider
+    );
+  }
+
+  // There is at least one event emitted during the rewards interval
+  const applicationEvents = intervalEvents.filter((obj) => {
+    return obj.args.application == application.address;
+  });
+
+  return await authorizationForRewardsInterval(
+    applicationEvents,
+    startRewardsBlock,
+    endRewardsBlock
+  );
+}
+
+async function filterEventsByApplications(events: any[]) {
+  const filteredEvents = new Array();
+  for (let j = 0; j < events.length; j++) {
+    const event = events[j];
+    if (
+      event.args.application == RandomBeaconAddress ||
+      event.args.application == WalletRegistryAddress
+    ) {
+      filteredEvents.push(event);
+    }
+  }
+
+  return filteredEvents;
+}
+
+async function filterEventsByStakingProvider(
+  events: any[],
+  stakingProvider: string
+) {
+  const filteredEvents = new Array();
+  for (let j = 0; j < events.length; j++) {
+    const event = events[j];
+    if (event.args.stakingProvider == stakingProvider) {
+      filteredEvents.push(event);
+    }
+  }
+
+  return filteredEvents;
+}
+
+// Calculates the weighted authorization for rewards interval based on events.
+// The general idea of weighted rewards calculation describes the following example.
+// Please note that this example operates on dates for simplicity purposes,
+// however the actual calculation is based on block numbers.
+// Ex.
+// events:         ^     ^      *    *   ^
+// timeline:  -|---|-----|------|----|---|--------|-->
+//         Sep 0   3     8      14  18   22       30
+// where: '^' denotes increase in authorization
+//        '*' denotes decrease in authorization
+//         0 -> Sep 1 at 00:00:00
+// event authorizations:
+//  Sep 0 - 3 from 100k to 110k
+//  Sep 3 - 8 from 110k to 135k
+//  Sep 8 - 14 from 135k to 120k
+//  Sep 14 - 18 from 120k to 100k
+//  Sep 18 - 30 constant 100k (last sub-interval)
+// authorization = (3-0)/30*100 + (8-3)/30*110 + (14-8)/30*135
+//               + (18-14)/30*120 + (30-18)/30*100
+async function authorizationForRewardsInterval(
+  events: any[],
+  startRewardsBlock: number,
+  endRewardsBlock: number
+) {
+  if (events.length == 0) {
+    return BigNumber.from("0");
+  }
+
+  let authorization = BigNumber.from("0");
+  const deltaRewardsBlock = endRewardsBlock - startRewardsBlock;
+  // ascending order
+  events.sort(function (a, b) {
+    return a.blockNumber < b.blockNumber
+      ? -1
+      : a.blockNumber == b.blockNumber
+      ? 0
+      : 1;
+  });
+
+  let tmpBlock = startRewardsBlock; // prev tmp block
+  let event;
+  for (let i = 0; i < events.length; i++) {
+    event = events[i];
+    const coefficientStr = (
+      ((event.blockNumber - tmpBlock) / deltaRewardsBlock) *
+      PRECISION
+    ).toString();
+    const coefficient = parseInt(coefficientStr);
```

Comment:
can we bind this directly? I don't see us using `coefficientStr` anywhere else:

```js
    const coefficient = ((event.blockNumber - tmpBlock) / deltaRewardsBlock) *
      PRECISION
```

Same with below

Reply:
The idea was to make a floating number like `42.0` to an integer `42` because BigNumber doesn't support floating numbers. I used now `Math.floor` instead of the previous approach which I agree was over-engineered. As part of [4505ecf](https://github.com/keep-network/keep-core/pull/3346/commits/4505ecf04eda7051cc0ddea2e336107d3c292fd2)